application_test.go

```diff
@@ -427,3 +427,20 @@ func TestApplicationConfiguration(t *testing.T) {
 	assert.True(t, ok)
 	assert.Equal(t, ErrCodeNotFound, apiErr.ErrCode)
 }
+
+func TestWaitOnApplication(t *testing.T) {
+	endpoint := newFakeMarathonEndpoint(t, nil)
+	defer endpoint.Close()
+	client := endpoint.Client.(*marathonClient)
+
+	c := make(chan bool)
+	err := client.waitOnApplication(fakeAppName, c)
+	assert.NoError(t, err)
+
+	c = make(chan bool)
+	go func() {
+		c <- true
+	}()
```

Comment:
I played with this function once more, and I wonder, is there a reason why custom timer function (`deadline`) is used there?
I ended up with this 

```go
func (r *marathonClient) WaitOnApplication(name string, timeout time.Duration) error {

// we have to run this for the first time
    if r.appExistAndRunning(name) {
        return nil
    }

    deadline := time.NewTimer(timeout).C
    ticker := time.NewTicker(time.Millisecond * 500).C
    for {
        select {
        case <-deadline:
            return ErrTimeoutError
        case <-ticker:
            if r.appExistAndRunning(name) {
                return nil
            }
        }
    }
    return nil
    }
func (r *marathonClient) appExistAndRunning(name string) bool {

    app, err := r.Application(name)
    if apiErr, ok := err.(*APIError); ok && apiErr.ErrCode == ErrCodeNotFound {
        return false
    }
    if err == nil && app.AllTaskRunning() {
        return true
    }
    return false}
```

with the test
```go
func TestWaitOnApplication(t *testing.T) {
    endpoint := newFakeMarathonEndpoint(t, nil)
    defer endpoint.Close()
    client := endpoint.Client.(*marathonClient)

    err := client.WaitOnApplication(fakeAppName, 1*time.Second)
    assert.NoError(t, err)

    err = client.WaitOnApplication("no_such_app", 1*time.Second)
    assert.IsType(t, err, ErrTimeoutError)
}
func TestAppExistAndRunning(t *testing.T) {
    endpoint := newFakeMarathonEndpoint(t, nil)
    defer endpoint.Close()
    client := endpoint.Client.(*marathonClient)
    assert.True(t, client.appExistAndRunning(fakeAppName))
    assert.False(t, client.appExistAndRunning("no_such_app"))
}
```

What do you think about it?